src/core/mempool.rs

```diff
@@ -284,6 +286,134 @@ impl MemPoolDB {
         })
     }
 
+    fn walk(&self, chainstate: &mut StacksChainState, tip_burn_header_hash: &BurnchainHeaderHash, tip_block_hash: &BlockHeaderHash, tip_height: u64) -> Result<Option<(BurnchainHeaderHash, BlockHeaderHash, u64, u64)>, ChainstateError> {
+        // Walk back to the next-highest
+        // ancestor of this tip, and see if we can include anything from there.
+        let next_height = MemPoolDB::get_previous_block_height(&self.db, tip_height)?.unwrap_or(0);
+        if next_height == 0 && tip_height == 0 {
+            // we're done -- tried every tx
+            debug!("Done scanning mempool -- at height 0");
+            return Ok(None);
+        }
+
+        let mut next_tips = MemPoolDB::get_chain_tips_at_height(&self.db, next_height)?;
+        if next_tips.len() == 0 {
+            // we're done -- no more chain tips
+            debug!("Done scanning mempool -- no chain tips at height {}", next_height);
+            return Ok(None);
+        }
+        
+        let ancestor_tip = {
+            let mut headers_tx = chainstate.headers_tx_begin()?;
+            // `next_height` is 1-indexed, since stacks blocks start at height 1.
+            // The headers DB, however, is 0-indexed, since it includes the genesis
+            // block header hash.  Account for this.
+            let next_header_height = next_height.saturating_sub(1);
```

Comment:
Just a heads-up -- in #1563, we eliminate `next_header_height` altogether, since the headers DB now starts at 0 (as it should have all along).  I fixed this bug in my tree, but you may need to do it here as well if #1563 merges first.